### PR TITLE
[fix #2] edge calculator

### LINKED ISSUE #2: StreamNet Transaction Weight Calculation
*Note*
Implement the weight calculator based on both vertex and edge information.
1) Start from the entry point, which is calculated from entryPointSelector.getEntryPoint, calculate every edges's length, which is induced from the difference in timestamp L(u,v) = (TS(u)-TS(v)). Then each edge's weight would be w(u,v) = L(u,v) / max(L(i,j)). 
2) calculate the vertex weight based on both vertex and edge info. Follow the below rule:
`set(u) = set(v)*w(u,v) + set(u)`
3) round the result to integer.

### Description
Provide the CumulativeWeightWithEdgeCalculator.calculate interface.

### Motivation
This feature is used for protection against the para-chain attack.

### Am I planning to do it myself with a PR?
Yes

### PR DESCRIPTION
**Pull request by stplaydog**

### CODE REVIEW

File: src/main/java/com/iota/iri/controllers/TransactionViewModel.java [hunk 1]
```diff
@@ -60,8 +60,8 @@
 
     private byte[] trits;
     public int weightMagnitude;
-
-    public static void fillMetadata(Tangle tangle, TransactionViewModel transactionViewModel) throws Exception {
+
```
Comment (stplaydog) [inline src/main/java/com/iota/iri/controllers/TransactionViewModel.java]: Please remove these redundant changes.

File: src/main/java/com/iota/iri/service/API.java [hunk 2]
```diff
@@ -680,8 +680,18 @@ public void storeTransactionsStatement(final List<String> trytes) throws Excepti
...
+            if(transactionViewModel.store(instance.tangle)) {
+                transactionViewModel.setArrivalTime(System.currentTimeMillis() / 1000L);
+                instance.transactionValidator.updateStatus(transactionViewModel);
+                transactionViewModel.updateSender("local");
+                transactionViewModel.update(instance.tangle, "sender");
+            }
+            // move store block from the for() function below --end
             elements.add(transactionViewModel);
         }
+        /*
```
Comment (stplaydog) [inline src/main/java/com/iota/iri/service/API.java]: Please just remove these lines, as you can see, elements can also be removed right?

File: src/main/java/com/iota/iri/service/tipselection/impl/CumulativeWeightCalculator.java [hunk 3]
```diff
@@ -119,14 +136,13 @@ private Hash getAndRemoveApprover(Collection<Hash> appHashes) {
...
         Hash branchHash = transactionViewModel.getBranchTransactionHash();
-
         Set<HashId> trunkApprovers = createApprovers(txHashToApprovers, txHash, approvers, trunkHash);
         txHashToApprovers.put(trunkHash, trunkApprovers);
         Set<HashId> branchApprovers = createApprovers(txHashToApprovers, txHash, approvers, branchHash);
         txHashToApprovers.put(branchHash, branchApprovers);
 
-        txHashToApprovers.remove(txHash);
-
+        //???? 这行代码完全没有起作用
```
Comment (stplaydog) [inline src/main/java/com/iota/iri/service/tipselection/impl/CumulativeWeightCalculator.java]: Don't write this kind of comments in production code.

File: src/test/java/com/iota/iri/service/tipselection/impl/CumulativeWeightWithEdgeCalculatorTest.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        transaction4.setArrivalTime(arriveTime);
+        transaction4.store(tangle);
+        arriveTime += 400;
+        transaction5.setArrivalTime(arriveTime);
+        transaction5.store(tangle);
+        UnIterableMap<HashId, Integer> txToCw = cumulativeWeightWithEdgeCalculator.calculate(transaction.getHash());
+
+        
+        Assert.assertEquals(String.format(TX_CUMULATIVE_WEIGHT_IS_NOT_AS_EXPECTED_FORMAT, 5),
+                6, txToCw.get(transaction5.getHash()).intValue());
```
Comment (stplaydog) [inline src/test/java/com/iota/iri/service/tipselection/impl/CumulativeWeightWithEdgeCalculatorTest.java]: This test is actually doing nothing?

**Review by stplaydog — COMMENTED** [review]
Please also change the comment from Chinese to English, hopefully these lines of code will be noticed by the IOTA community.

File: src/main/java/com/iota/iri/service/tipselection/impl/CumulativeWeightCalculator.java [hunk 5]
```diff
@@ -1,5 +1,6 @@
 package com.iota.iri.service.tipselection.impl;
```
Comment (stplaydog) [inline src/main/java/com/iota/iri/service/tipselection/impl/CumulativeWeightCalculator.java]: I think it's better not to change anything in this file.

File: src/main/java/com/iota/iri/service/tipselection/impl/CumulativeWeightWithEdgeCalculator.java [hunk 6]
```diff
@@ -16,15 +26,140 @@
...
+		super(tangle);
+	}
 
-    @Override
-    public UnIterableMap<HashId, Integer> calculate(Hash entryPoint) throws Exception {
-        return super.calculate(entryPoint);
-    }
-}
+	@Override
+	public UnIterableMap<HashId, Integer> calculate(Hash entryPoint) throws Exception {
```
Comment (stplaydog) [inline src/main/java/com/iota/iri/service/tipselection/impl/CumulativeWeightWithEdgeCalculator.java:30]: This function could be reused in CumulativeWeightCalculator.java?

File: src/main/java/com/iota/iri/service/tipselection/impl/CumulativeWeightWithEdgeCalculator.java [hunk 7]
```diff
@@ -16,15 +26,140 @@
...
+	}
+
+	private Hash getAndRemoveApprover(Collection<Hash> appHashes) {
+		Iterator<Hash> hashIterator = appHashes.iterator();
+		Hash txApp = hashIterator.next();
+		hashIterator.remove();
+		return txApp;
+	}
 
+	private Collection<Hash> getTxDirectApproversHashes(Hash txHash, Map<Hash, Collection<Hash>> txToDirectApprovers)
```
Comment (stplaydog) [inline src/main/java/com/iota/iri/service/tipselection/impl/CumulativeWeightWithEdgeCalculator.java]: same

File: src/main/java/com/iota/iri/controllers/TransactionViewModel.java [hunk 8]
```diff
@@ -210,10 +210,30 @@ public boolean store(Tangle tangle) throws Exception {
...
+		trunk = getTrunkTransaction(tangle);
+		transaction.trunkTimeDifference = trunk.getArrivalTime() == 0 ? 0 : (transaction.arrivalTime - trunk.getArrivalTime());
+	}
+	
+	private void setBranchTimeDifference(Tangle tangle) throws Exception {
+		branch = getBranchTransaction(tangle);
+		transaction.branchTimeDifference = branch.getArrivalTime() == 0 ? 0 : (transaction.arrivalTime - branch.getArrivalTime());
+	}
+	
+    public long getTrunkTimeDifference() {
```
Comment (wunder3605) [inline src/main/java/com/iota/iri/controllers/TransactionViewModel.java]: Missing a space?

File: src/main/java/com/iota/iri/service/tipselection/impl/CumulativeWeightWithEdgeCalculator.java [hunk 9]
```diff
@@ -16,15 +26,140 @@
...
+	}
+
+	private UnIterableMap<HashId, Integer> calculateCwInOrder(LinkedHashSet<Hash> txsToRate) throws Exception {
+		// 计算最大时长以及初始化各边时长
+		long maxTimeDiff = 0l;
+		long timeDiff = 0l;
+		Iterator<Hash> txHashIterator = txsToRate.iterator();
+		Map<String, Long> edgeMap = new HashMap<String, Long>();
+		Map<Hash, Set<Hash>> approvers = new HashMap<Hash, Set<Hash>>();
+		while (txHashIterator.hasNext()) {
```
Comment (stplaydog) [inline src/main/java/com/iota/iri/service/tipselection/impl/CumulativeWeightWithEdgeCalculator.java:63]: Spare this part of the code to a separate function

File: src/main/java/com/iota/iri/service/tipselection/impl/CumulativeWeightWithEdgeCalculator.java [hunk 10]
```diff
@@ -1,30 +1,131 @@
 package com.iota.iri.service.tipselection.impl;
 
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.Map;
```
Comment (stplaydog) [inline src/main/java/com/iota/iri/service/tipselection/impl/CumulativeWeightWithEdgeCalculator.java]: util.*

**Review by wunder3605 — APPROVED** [review]
+1

**Pull request merged**